api/src/main/java/org/xwiki/apps/googleapps/internal/GoogleAppsAuthServiceImpl.java

```diff
@@ -0,0 +1,252 @@
+     * @param username the provided user-name
+     * @param password the provided password
+     * @param context the context describing the request
+     * @return a null Principal Object if the user hasn't been authenticated or a valid Principal Object if the user is
+     *         correctly authenticated
+     * @throws XWikiException if something goes wrong.
+     */
+    public Principal authenticate(String username, String password, XWikiContext context) throws XWikiException {
+        try {
+            log.info("GoogleApps authentificator - authenticate");
+
+            // case of a too early call or deactivated... can only count on local users
+            if (googleAppsManager == null || !googleAppsManager.isActive(context)) {
+                return super.authenticate(username, password, context);
+            }
+
+            HttpSession session = context.getRequest().getSession();
+            String xwikiUser = (String) session.getAttribute("googleappslogin");
+            log.info("xwikiUser from session : " + xwikiUser);
+            // get configuration for authentification with cookies
+
+            // authenticate user from cookie value
+            if (xwikiUser == null && googleAppsManager.useCookies() && googleAppsManager.authWithCookies()) {
+                log.info("Authenticate with cookie");
+                CookieAuthenticationPersistence cookieTools =
+                        componentManager.getInstance(CookieAuthenticationPersistence.class);
+                cookieTools.initialize(context, googleAppsManager.getConfigCookiesTTL());
+                String userCookie = cookieTools.retrieve();
+                if (userCookie != null) {
+                    log.info("Found user from cookie : " + userCookie);
+                    DocumentReference userDocRef = googleAppsManager.createUserReference(username);
+                    XWikiDocument userDoc = context.getWiki().getDocument(userDocRef, context);
+                    if (!userDoc.isNew())  {
+                        xwikiUser = userDocRef.getName();
+                    }
+                    log.info("xwikiUser from cookie : " + xwikiUser);
+                }
+            }
+            if (xwikiUser != null) {
+                if (!xwikiUser.startsWith(XWIKISPACE)) {
+                    xwikiUser = XWIKISPACE + xwikiUser;
+                }
+                log.info("Authenticating user " + xwikiUser);
+                return new SimplePrincipal(xwikiUser);
+            } else {
+                log.info("use default authenticate method for user : " + username);
+                return super.authenticate(username, password, context);
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw new XWikiException("Trouble at authenticating", e);
+        }
+    }
+
+    private Pattern logoutRequestMatcher;
+
+    @Inject
+    @Named("xwikicfg")
+    private Provider<ConfigurationSource> xwikicfgProvider;
```

Comment:
Class fields should be listed before the class methods.

Reply:
done